README.md
# Infinite Resources

This is a mod for the game [Factorio](https://www.factorio.com/). It can be downloaded from the [Factorio Mod Portal](https://mods.factorio.com/mod/sonaxaton-nuclear-fuel-krastorio-compat).

## Description

A small mod to re-balance GotLag's [Nuclear Fuel](https://mods.factorio.com/mod/Nuclear%20Fuel) mod when used with the overhaul mod [Krastorio 2](https://mods.factorio.com/mod/Krastorio2). Both mods make changes to nuclear processing recipes and so without this mod you end up with a completely imbalanced fuel reprocessing cycle. K2 also changes the fuel values of uranium fuel cells, so this mod updates the breeder fuel cells from NF to have the correct relative energy capacity.

Full list of changes:

* Re-balances Nuclear Fuel Reprocessing to consume/produce the same relative number of uranium fuel cells and uranium-238 as NF does with vanilla, based on the fuel cycle setting. For the Classic cycle, this multiplies the K2 recipe's fuel cells and uranium-238 by 4. For the Alternative cycle, this multiplies fuel cells by 2 and uranium-238 by 4/3. The byproducts added by K2 (stone and tritium) and NF (plutonium) are not adjusted.
* Updates MOX fuel to produce 5x fewer uranium fuel cells since in K2 fuel cells are much more energy-rich. This aligns with the fact that K2 divides the vanilla number of fuel cells in Nuclear Fuel Reprocessing by 5.
* Updates the breeder fuel cell recipe to produce 5x fewer cells as well.
* Updates the breeder fuel cell reprocessing recipe to consume 5x fewer cells as well as produce more uranium-235. Without K2, this recipe produces enough uranium-235 to make the same number of uranium fuel cells as breeder cells that were reprocessed. With K2 and this mod, the amount is adjusted to maintain this ratio. The updated recipe will consume 2 used breeder cells and produce 4 uranium-235, which is enough to make 2 uranium fuel cells. If [Schall Uranium Processing](https://mods.factorio.com/mod/SchallUraniumProcessing) is installed, this amount is adjusted for the changes it makes to be 8 low-enriched uranium.
* Updates the NF recipe to make nuclear fuel from rocket fuel and plutonium to use less plutonium since plutonium is harder to make. With the change you get 4x more nuclear fuel for the same amount of plutonium.
* Updates the NF recipe to make atomic bombs from plutonium to require a heavy rocket like the normal recipe does.
* Updates the energy capacity of breeder fuel cells to be the same ratio to K2's uranium cells as they are in vanilla (half).
* Updates the stack sizes of the items added by NF (plutonium, breeder fuel cells, used breeder fuel cells) with the same updates to similar items made by K2.
* Removes the Kovarex Enrichment Process technology as a prerequisite from any other technology since NF disables it. For K2 this includes the Fusion Energy technology, but it will work for any other technologies added by other mods as well.


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.2.1
Date: 2020-08-14
  Minor Changes:
    - Factorio 1.0!
---------------------------------------------------------------------------------------------------
Version: 0.2.0
Date: 2019-08-02
  Changes:
    - Update NF's Atomic Bomb from Plutonium recipe to include Heavy Rocket from K2.
---------------------------------------------------------------------------------------------------
Version: 0.1.0
Date: 2019-07-28
  First release!


data-final-fixes.lua
-- RECIPES

local nuclear_fuel_cycle_type = settings.startup['nuclear-fuel-cycle-type'].value
local nuclear_fuel_cycle_ingredient_multiplier
local nuclear_fuel_cycle_product_multiplier
if nuclear_fuel_cycle_type == 'classic' then
  nuclear_fuel_cycle_ingredient_multiplier = 4
  nuclear_fuel_cycle_product_multiplier = 4
else
  nuclear_fuel_cycle_ingredient_multiplier = 2
  nuclear_fuel_cycle_product_multiplier = 4/3
end

-- multiply K2 nuclear fuel reprocessing recipe core products
-- keep Pu added by NF and energy already multiplied by NF
-- also don't change the stone and tritium byproducts added by K2
-- the end result is the K2 recipe multiplied by NF's multiplier with the Pu added by NF
krastorio.recipes.multiplyIngredient(
  'nuclear-fuel-reprocessing',
  'used-up-uranium-fuel-cell',
  nuclear_fuel_cycle_ingredient_multiplier)
krastorio.recipes.multiplyProduct(
  'nuclear-fuel-reprocessing',
  'uranium-238',
  nuclear_fuel_cycle_product_multiplier)
data.raw.recipe['nuclear-fuel-reprocessing'].main_product = nil

-- update NF MOX fuel cell recipe to produce the same relative amount of fuel cells per Pu as it does when just used with vanilla
-- this is balanced because fuel cells are more energy-rich in K2
-- K2 reduces fuel cells in nuclear fuel reprocessing by a factor of 5
-- so we want to do the same with the fuel cells in MOX
-- also, K2 replaces the 1 iron plate per fuel cell in the fuel cell recipe with 1 steel plate, so we do the same here
krastorio.recipes.replaceIngredient(
  'mox-fuel-cell',
  'iron-plate',
  {'steel-plate', 2})
krastorio.recipes.replaceProduct(
  'mox-fuel-cell',
  'uranium-fuel-cell',
  {'uranium-fuel-cell', 2})

-- update NF breeder fuel cell recipe, exact same logic as MOX fuel cell recipe
krastorio.recipes.replaceIngredient(
  'breeder-fuel-cell',
  'iron-plate',
  {'steel-plate', 2})
krastorio.recipes.replaceProduct(
  'breeder-fuel-cell',
  'breeder-fuel-cell',
  {'breeder-fuel-cell', 2})

-- update NF breeder fuel reprocessing recipe to use 5x less cells for the same amount of Pu
-- also make it produce more U235
-- with vanilla, breeder reprocessing produces enough U235 to make the same number of fuel cells as breeder fuel cells that were reprocessed
-- so with K2 we want the same ratio: 4 U235 makes 2 fuel cells
-- if Schall's Uranium Processing AND K2 are installed, this becomes 8 LEU (with just Schall it's 4 but K2's compat script multiplies it by 2) to make 2 fuel cells
krastorio.recipes.replaceIngredient(
  'breeder-fuel-reprocessing',
  'used-up-breeder-fuel-cell',
  {'used-up-breeder-fuel-cell', 2})
local breeder_reprocessing_product
if mods['SchallUraniumProcessing'] then
  breeder_reprocessing_product = {'uranium-low-enriched', 8}
else
  breeder_reprocessing_product = {'uranium-235', 4}
end
krastorio.recipes.replaceProduct(
  'breeder-fuel-reprocessing',
  'uranium-235',
  breeder_reprocessing_product)

-- update NF nuclear fuel from Pu to produce more rocket fuel since Pu is harder to produce
krastorio.recipes.multiplyIngredient(
  'nuclear-fuel-pu',
  'rocket-fuel',
  4)
krastorio.recipes.multiplyProduct(
  'nuclear-fuel-pu',
  'nuclear-fuel',
  4)

-- update atomic bomb from Pu recipe to include heavy rocket as K2 does for the normal recipe
krastorio.recipes.addIngredient(
  'atomic-bomb-pu',
  {'heavy-rocket', 1})


-- STACK SIZES

-- update stack sizes of items added by NF that should have been modified by K2
local kr_stack_size_value = krastorio.general.getSafeSettingValue('kr-stack-size')
if kr_stack_size_value and kr_stack_size_value ~= 'No changes' then
  data.raw.item['plutonium'].stack_size = kr_stack_size_value
  data.raw.item['breeder-fuel-cell'].stack_size = 10
  data.raw.item['used-up-breeder-fuel-cell'].stack_size = 10
end


-- FUEL VALUES

-- update breeder fuel cell value to keep the same ratio it had to vanilla fuel cells (half)
data.raw.item['breeder-fuel-cell'].fuel_value = '25GJ'


-- TECHNOLOGY

-- NF disables the Kovarex Enrichment Process technology completely
-- but some K2 techs depend on it
-- so remove those dependencies
for technology_name, _ in pairs(data.raw.technology) do
  if krastorio.technologies.removePrerequisite(
    technology_name,
    'kovarex-enrichment-process') then
    log('removed kovarex-enrichment-process prerequisite from '..technology_name)
  end
end
